server/routerlicious/packages/services/src/throttler.ts

```diff
@@ -0,0 +1,113 @@
+        }
+
+        // Exit early if already throttled and no chance of being unthrottled
+        const timeUntilNotThrottled = this.getTimeUntilNotThrottledInMs(requestMetric, now);
+        if (timeUntilNotThrottled > 0) {
+            requestMetric.retryAfterInMs = timeUntilNotThrottled;
+            // update stored request metric with new retry duration
+            await this.throttleManager.setRequestMetric(id, requestType, requestMetric);
+            return this.convertRequestMetricsToThrottlerResponse(requestMetric);
+        }
+
+        // replenish "tokens" if possible
+        const amountToReplenish = this.getAmountToReplenishOnCooldown(requestMetric, now);
+        if (amountToReplenish > 0) {
+            requestMetric.count += amountToReplenish;
+            requestMetric.lastCoolDownAt = now;
+        }
+
+        // adjust "tokens" based on given count
+        requestMetric.count -= count;
+
+        // throttle if "token bucket" is empty
+        const newTimeUntilNotThrottled = this.getTimeUntilNotThrottledInMs(requestMetric, now);
+        if (newTimeUntilNotThrottled > 0) {
+            requestMetric.throttleStatus = true;
+            requestMetric.throttleReason = `count exceeded by ${Math.abs(requestMetric.count)}`;
+            requestMetric.retryAfterInMs = newTimeUntilNotThrottled;
+        } else {
+            requestMetric.throttleStatus = false;
+            requestMetric.throttleReason = "";
+            requestMetric.retryAfterInMs = 0;
+        }
+
+        // update stored request metric
+        await this.throttleManager.setRequestMetric(id, requestType, requestMetric);
+
+        return this.convertRequestMetricsToThrottlerResponse(requestMetric);
+    }
+
+    public async getThrottleStatus(id: string, requestType: ThrottlerRequestType): Promise<IThrottlerResponse> {
+        const requestMetric = await this.throttleManager.getRequestMetric(id, requestType);
+        return this.convertRequestMetricsToThrottlerResponse(requestMetric);
+    }
+
+    private convertRequestMetricsToThrottlerResponse(requestMetric: IRequestMetrics): IThrottlerResponse {
+        return {
+            throttleStatus: requestMetric.throttleStatus,
+            throttleReason: requestMetric.throttleReason,
+            retryAfterInMs: requestMetric.retryAfterInMs,
+        };
+    }
+
+    private getAmountToReplenishOnCooldown(requestMetric: IRequestMetrics, now: number): number {
+        const timeSinceLastCooldownInMs = now - requestMetric.lastCoolDownAt;
+        // replenish "tokens" at most once per minCooldownInterval
+        if (timeSinceLastCooldownInMs > this.minCooldownIntervalInMs) {
+            return Math.floor(timeSinceLastCooldownInMs / this.requestRate);
+        }
+        return 0;
+    }
+
+    private getTimeUntilNotThrottledInMs(requestMetric: IRequestMetrics, now: number): number {
+        const debt = 0 - requestMetric.count;
+        if (debt <= 0) {
+            return 0;
+        }
```

Comment:
I'm not sure if I understand this condition. Could you please explain this part?

Reply:
Following up on this: the short-circuit here is actually required, otherwise `Math.max` on line `114` will choose `timeUntilNextCooldown` when `timeUntilDebtRepleneshed < 0`, so requests that should be un-throttled become throttled for cooldown duration. Would moving this conditional below line 112, and returning `timeUntilDebtReplenished` be better?